### DISCUSSION TITLE
How is it different from Linaria?

### DISCUSSION THREAD

**Discussion by zaguiini** [microsoft/griffel#60] [Q&A] [answered]
Hi! Great project!

I'm trying to understand how this project differs from existing solutions, like https://github.com/callstack/linaria for example?

**Comment by layershifter** [comment, marked as answer]
We are building library that works, scales and is performant for UI libraries/frameworks & applications. And this affects its design and goals. We use Atomic CSS to solve [specificity problems](https://www.w3.org/TR/selectors/#specificity) and [order of appearance](https://www.w3.org/TR/css-cascade-3/#cascade-order). This is the key difference/similarity to other libraries.

Here is a sample CodeSandbox (https://codesandbox.io/s/order-of-appearance-puzzle-svwi8) that shows why order of appearance matters 😉 

In terms of competitors the closest ones is [Compiled](https://github.com/atlassian-labs/compiled) and [Stylex](https://www.youtube.com/watch?v=ur-sGzUWId4). It's hard to compare with the second, but compared to Compiled we have significantly better runtime performance:
![MicrosoftTeams-image](https://user-images.githubusercontent.com/14183168/156006218-cc8564ff-188d-4336-aeb8-81e88df55f62.png)

(👆 _I plan to publish this benchmark later somewhere, the same as in  [react-native-web](https://github.com/necolas/react-native-web/tree/master/packages/benchmarks) & in [fela](https://github.com/necolas/react-native-web/tree/master/packages/benchmarks)_ but contains more libraries)

We don't have similar metrics as Meta with Stylex, but once CSS extraction #53 will be supported we expect to see the same bundle size curve: more styles => less bundle size.

# vs. Linaria

Linaria overall is a perfect library we even reuse [their module evaluation](https://github.com/microsoft/griffel/blob/cfee239c4fd6f2b5bfe075187d72dfd2635e9dda/packages/babel-preset/src/utils/evaluatePathsInVM.ts) for our build time transforms ❤️ 

## From API

- Griffel has syntax based on JavaScript objects: Linaria uses template strings with CSS
  ```tsx
  const useHeaderClasses = makeStyles({
    root: {
      textTransform: "uppercase",
      fontFamily: "Segoe UI"
    }
  });
  // vs.
  const header = css`
    text-transform: uppercase;
    font-family: Segoe UI;
  `;
  ```
- Griffel uses Atomic CSS: Linaria uses monolithic classes
- Griffel **works in runtime**, it's zero config like other CSS-in-JS (Emotion, Styled components). This helps customers to use it instantly with complicated setup

## From implementation perspective

There are other competitors on this field that are similar to Linaria (for example, [vanilla-extract](https://github.com/seek-oss/vanilla-extract)), but in fact they are more or less "wrappers" around [CSS modules](https://github.com/css-modules/css-modules). There is nothing wrong with that approach, but it does not scale for reusable libraries well.

Let's take [an example from Linaria docs](https://codesandbox.io/s/linaria-demo-forked-lf3wzs?file=/src/App.js:102-322),  there is a basic button `Button` and `LargeButton` that uses `styled` API:

![image](https://user-images.githubusercontent.com/14183168/155989473-3a2ad18b-9cec-4a13-864c-8fa9a38d97e2.png)

We can notice that the library uses a trick with increased specificity to ensure that styles from `LargeButton` will win. This solves the problem with [order of appearance](https://www.w3.org/TR/css-cascade-3/#cascade-order) for first party components. Another basic pattern in Linaria is [to refer components in styles](https://github.com/callstack/linaria/blob/master/docs/BASICS.md#referring-to-another-component-or-class-name):

```ts
const Article = styled.article`
  ${LargeButton} { color: white; }
`;
```

It's a moment when order of appearance strikes with async imports and increased specificity does not help:

![image](https://user-images.githubusercontent.com/14183168/156014705-a32696f9-6797-4366-8d3f-1a48547cc6d5.png)
![image](https://user-images.githubusercontent.com/14183168/156014643-d0130b2e-7860-4ae1-bdda-721f7c6e763c.png)

Although Griffel does not anything with specificity for composition (classes merging and deduplication happens in `mergeClasses()` function) it also does not depend so much on order of appearance (_there are still might be rare edge cases_ 💥 ). The result is correct:

![image](https://user-images.githubusercontent.com/14183168/156014769-51c496c9-56ef-4229-900e-a2e0b2697d62.png)

- Linaria CodeSandbox: https://codesandbox.io/s/linaria-demo-forked-lf3wzs
- Griffel CodeSandbox: https://codesandbox.io/s/hardcore-sea-61zvfk

While the order of appearance looks like an edge case that will be more common soon with increasing usage of `React.Suspense`.

P.S. There was a recent check in of Atomic CSS approach (https://github.com/callstack/linaria/pull/867), but I didn't manage to check & compare.
